extensions/binding/org.eclipse.smarthome.binding.homematic/src/main/java/org/eclipse/smarthome/binding/homematic/internal/communicator/virtual/ButtonVirtualDatapointHandler.java

```diff
@@ -0,0 +1,89 @@
+ *
+ * See the NOTICE file(s) distributed with this work for additional
+ * information regarding copyright ownership.
+ *
+ * This program and the accompanying materials are made available under the
+ * terms of the Eclipse Public License 2.0 which is available at
+ * http://www.eclipse.org/legal/epl-2.0
+ *
+ * SPDX-License-Identifier: EPL-2.0
+ */
+package org.eclipse.smarthome.binding.homematic.internal.communicator.virtual;
+
+import static org.eclipse.smarthome.binding.homematic.internal.misc.HomematicConstants.VIRTUAL_DATAPOINT_NAME_BUTTON;
+
+import org.apache.commons.lang.StringUtils;
+import org.eclipse.smarthome.binding.homematic.internal.misc.MiscUtils;
+import org.eclipse.smarthome.binding.homematic.internal.model.HmChannel;
+import org.eclipse.smarthome.binding.homematic.internal.model.HmDatapoint;
+import org.eclipse.smarthome.binding.homematic.internal.model.HmDevice;
+import org.eclipse.smarthome.binding.homematic.internal.model.HmValueType;
+import org.eclipse.smarthome.core.thing.CommonTriggerEvents;
+import org.eclipse.smarthome.core.thing.DefaultSystemChannelTypeProvider;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+/**
+ * A virtual String datapoint which adds a BUTTON datapoint. It will forward key events to the
+ * system channel {@link DefaultSystemChannelTypeProvider#SYSTEM_BUTTON}.
+ *
+ * @author Michael Reitler - Initial contribution
+ */
+public class ButtonVirtualDatapointHandler extends AbstractVirtualDatapointHandler {
+    private final Logger logger = LoggerFactory.getLogger(ButtonVirtualDatapointHandler.class);
+
+    @Override
+    public String getName() {
+        return VIRTUAL_DATAPOINT_NAME_BUTTON;
+    }
+
+    @Override
+    public void initialize(HmDevice device) {
+        for (HmChannel channel : device.getChannels()) {
+            if (channel.hasPressDatapoint()) {
+                HmDatapoint dp = addDatapoint(device, channel.getNumber(), getName(), HmValueType.STRING, null, false);
+                dp.setTrigger(true);
+                dp.setOptions(new String[] { CommonTriggerEvents.SHORT_PRESSED, CommonTriggerEvents.LONG_PRESSED,
+                        CommonTriggerEvents.DOUBLE_PRESSED });
+            }
+        }
+    }
+
+    @Override
+    public boolean canHandleEvent(HmDatapoint dp) {
+        return dp.isPressDatapoint();
+    }
+
+    @Override
+    public void handleEvent(VirtualGateway gateway, HmDatapoint dp) {
+        HmDatapoint vdp = getVirtualDatapoint(dp.getChannel());
+        if (MiscUtils.isTrueValue(dp.getValue())) {
+            String pressType = StringUtils.substringAfter(dp.getName(), "_");
+            switch (pressType) {
+                case "SHORT":
+                    if (dp.getValue() == null || !dp.getValue().equals(dp.getPreviousValue())) {
+                        vdp.setValue(CommonTriggerEvents.SHORT_PRESSED);
+                    } else {
+                        // two (or more) PRESS_SHORT events were received
```

Comment:
Would this not also be true for two successive LONG_PRESSED events?

Reply:
The variable `dp` contains the original (non virtual) datapoint. This branch of the switch-case is only used for the `PRESS_SHORT` datapoint. Therefore it will not trigger on two successive `LONG_PRESSED` events

Reply:
Of course you're right; I missed that the original datapoint is used here.